### ISSUE TITLE
Why the output of model are different when setting `model.train()` and `model.eval()`? 

### ISSUE THREAD

**Issue by FalongFan** [ultralytics/yolov5#9318]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

I wonder what the output of Detect layer in training process, so I print out the output in when training. And I got the following result: 
```
tensor([-1.29590, -0.77002,  1.91113, -0.22046, -7.86328, -5.83984, -6.22266, -5.60547, -4.32812, -4.82031, -4.87500, -4.97266, -5.09766, -5.89844, -5.83203, -4.17969, -3.93750, -3.77539, -3.36328, -5.54688, -5.73047, -3.60156, -4.04297, -4.04688, -4.71094, -3.05078, -6.41406, -3.10938, -5.67969, -6.20312, -4.46094,
        -5.03906, -4.37109, -5.04297, -5.97656, -3.32031, -3.77148, -6.80859, -4.66016, -5.41797, -4.87500, -7.28516, -4.10547, -4.90625, -4.34766, -5.01953, -3.41016, -5.53125, -4.72266, -3.53516, -6.53125, -4.80469, -5.32422, -4.55859, -6.81250, -4.55078, -5.09375, -5.68750, -3.76562, -5.59375, -2.73828, -4.58594,
        -6.15625, -4.07031, -4.50391, -4.81250, -5.16797, -4.42578, -4.86328, -4.51172, -5.90625, -2.58789, -5.75391, -6.26953, -4.85938, -5.51953, -5.19141, -4.64062, -5.68359, -5.18750, -6.80469, -3.74219, -4.04688, -4.26562, -5.98047, -5.25391, -3.38477], device='cuda:0', dtype=torch.float16,
       grad_fn=<SliceBackward>)
```
Here I have 82 classes, training with `yolov5s.yaml`, and `pred[0][0, 0, 0, 0, :]` are printed. The codes are as below in `train.py`. 
```
...
             with torch.cuda.amp.autocast(amp):
                pred = model(imgs)  # forward
                print("======================[ pred[0][0, 0, 0, 0, :] ]=========================", pred[0][0, 0, 0, 0, :])
                loss, loss_items = compute_loss(pred, targets.to(device))  # loss scaled by batch_size
                if RANK != -1:
                    loss *= WORLD_SIZE  # gradient averaged between devices in DDP mode
                if opt.quad:
                    loss *= 4.
...
```
Why the pred are so strange with so many negative number? What I understand is that `pred[0][0, 0, 0, 0, :4]` are `xywh`, `pred[0][0, 0, 0, 0, 4]` is confidence and the rest of them are probabilities of all classes. Why probabilities are negative number? 

And, I did another experiment. I printed the predicted result in `detect.py`, and the results seems reasonable as what I understand. The code are as below in `detect.py`: 
```
...
         with dt[1]:
            visualize = increment_path(save_dir / Path(path).stem, mkdir=True) if visualize else False
            pred = model(im, augment=augment, visualize=visualize)
            print("=========================[ how pred looks like: pred.shape, pred[0, :2, :]]==========================", pred.shape, '\n', pred[0, :2, :])
...
```
And the output are: 
```
=========================[ how pred looks like: pred.shape, pred[0, :2, :]]========================== torch.Size([1, 18900, 85]) 
 tensor([[3.18788e+00, 4.33523e+00, 7.56535e+00, 1.14042e+01, 1.77960e-05, 1.06153e-01, 3.70127e-03, 2.39290e-02, 1.58053e-03, 1.55699e-03, 2.53002e-03, 2.97853e-03, 4.17790e-03, 3.21301e-03, 3.90965e-01, 2.86851e-03, 1.21180e-02, 6.11151e-03, 3.80492e-03, 3.86950e-02, 1.68551e-03, 1.07855e-03, 2.52746e-03, 3.21454e-03,
         1.27151e-03, 7.50361e-04, 4.35888e-04, 9.24049e-04, 1.55080e-03, 2.93206e-03, 1.15144e-02, 8.20965e-03, 3.54706e-03, 2.02936e-03, 9.70449e-04, 1.31942e-03, 6.55149e-04, 8.24752e-03, 4.51994e-03, 9.87016e-04, 7.88059e-04, 1.03882e-03, 1.22463e-03, 1.44330e-03, 2.17974e-02, 4.64432e-03, 1.12683e-02, 1.55740e-03,
         4.56563e-03, 2.68625e-03, 2.75892e-03, 1.66861e-02, 5.71842e-03, 7.21947e-04, 2.40221e-02, 1.58768e-03, 1.91814e-03, 6.97816e-04, 9.23993e-04, 2.27841e-03, 1.75540e-03, 1.73401e-02, 7.78530e-04, 3.75428e-02, 5.84692e-04, 2.12386e-03, 9.64389e-04, 3.95899e-03, 4.66836e-04, 8.51932e-04, 3.10042e-03, 6.39946e-04,
         5.06410e-03, 1.14314e-03, 8.16540e-04, 6.02816e-04, 1.46040e-03, 1.13674e-03, 1.46960e-02, 3.61738e-02, 6.60286e-03, 1.61258e-03, 2.28885e-03, 8.22704e-04, 1.32624e-03],
        [1.07240e+01, 4.57864e+00, 2.16118e+01, 9.27079e+00, 3.65370e-06, 1.33446e-01, 5.37072e-03, 5.04515e-02, 2.28957e-03, 2.29464e-03, 3.01200e-03, 3.69618e-03, 5.84781e-03, 5.37135e-03, 2.98573e-01, 2.15376e-03, 9.60508e-03, 4.80116e-03, 7.19322e-03, 4.64228e-02, 1.72779e-03, 1.36050e-03, 2.82397e-03, 4.67706e-03,
         1.87259e-03, 9.67565e-04, 6.44071e-04, 1.17832e-03, 1.61000e-03, 4.03842e-03, 1.33630e-02, 7.59444e-03, 4.04344e-03, 2.43891e-03, 1.43730e-03, 2.18082e-03, 8.87710e-04, 8.35485e-03, 4.99952e-03, 1.34363e-03, 1.16384e-03, 1.62665e-03, 1.93011e-03, 1.51906e-03, 1.33452e-02, 5.53018e-03, 1.33639e-02, 2.42505e-03,
         4.59119e-03, 3.55404e-03, 4.35506e-03, 1.04603e-02, 5.63497e-03, 1.01369e-03, 2.12796e-02, 1.88550e-03, 3.17702e-03, 1.05144e-03, 1.39801e-03, 4.10082e-03, 2.30359e-03, 2.69002e-02, 1.05125e-03, 2.35041e-02, 8.82470e-04, 4.29056e-03, 9.99579e-04, 3.74811e-03, 6.01923e-04, 1.18266e-03, 4.68550e-03, 9.61150e-04,
         7.37224e-03, 1.50084e-03, 1.14980e-03, 8.29530e-04, 2.85780e-03, 1.45193e-03, 1.64162e-02, 3.68443e-02, 4.56082e-03, 1.61329e-03, 2.61433e-03, 9.71655e-04, 1.43619e-03]], device='cuda:0')
```
Here in inference, I have 80 classes, and original weights of `yolov5s.pt`. The last dimension of `pred` is 85 with `xywh+confidence+probabilities of all classes`, and seems reasonable. 
So why the output of pred are not the same in training and detecting processes? 

Next, I have done another experiment, that is change the way the model built in detecting process. The explicit operation is channge the codes as below in `attempt_load` function in `experimental.py`, simply changed `.eval()` to `.train()`. 
```
...
def attempt_load(weights, device=None, inplace=True, fuse=True):
    # Loads an ensemble of models weights=[a,b,c] or a single model weights=[a] or weights=a
    from models.yolo import Detect, Model

    model = Ensemble()
    for w in weights if isinstance(weights, list) else [weights]:
        ckpt = torch.load(attempt_download(w), map_location='cpu')  # load
        ckpt = (ckpt.get('ema') or ckpt['model']).to(device).float()  # FP32 model
        # Model compatibility updates
        if not hasattr(ckpt, 'stride'):
            ckpt.stride = torch.tensor([32.])
        if hasattr(ckpt, 'names') and isinstance(ckpt.names, (list, tuple)):
            ckpt.names = dict(enumerate(ckpt.names))  # convert to dict
        # model.append(ckpt.fuse().eval() if fuse and hasattr(ckpt, 'fuse') else ckpt.eval())  # model in eval mode
        model.append(ckpt.fuse().train() if fuse and hasattr(ckpt, 'fuse') else ckpt.train())  # model in train mode
...
```
And I got the SAME results with training process!! As below: 
```
===============================[ len(pred), pred[0].shape, pred[1].shape, pred[2].shape, pred[0][0, 0, 0, 0, :] ]=============================== 3 torch.Size([1, 3, 80, 60, 85]) torch.Size([1, 3, 40, 30, 85]) torch.Size([1, 3, 20, 15, 85]) 
 tensor([ -0.20373,   0.08386,  -0.26191,  -0.12694, -10.93652,  -2.13065,  -5.59537,  -3.70845,  -6.44842,  -6.46344,  -5.97700,  -5.81334,  -5.47376,  -5.73733,  -0.44326,  -5.85109,  -4.40087,  -5.09145,  -5.56765,  -3.21258,  -6.38400,  -6.83106,  -5.97801,  -5.73685,  -6.66628,  -7.19421,  -7.73769,  -6.98582,
         -6.46743,  -5.82912,  -4.45258,  -4.79420,  -5.63808,  -6.19800,  -6.93678,  -6.62925,  -7.32999,  -4.78956,  -5.39473,  -6.91984,  -7.14515,  -6.86863,  -6.70389,  -6.53938,  -3.80393,  -5.36745,  -4.47443,  -6.46318,  -5.38462,  -5.91692,  -5.89015,  -4.07635,  -5.15833,  -7.23284,  -3.70447,  -6.44389,
         -6.25448,  -7.26686,  -6.98588,  -6.08200,  -6.34330,  -4.03724,  -7.15732,  -3.24401,  -7.44384,  -6.15239,  -6.94305,  -5.52780,  -7.66907,  -7.06715,  -5.77311,  -7.35349,  -5.28050,  -6.77283,  -7.10962,  -7.41330,  -6.52759,  -6.77846,  -4.20537,  -3.28258,  -5.01363,  -6.42831,  -6.07741,  -7.10209,
         -6.62408], device='cuda:0')
```
SO, my question is: why change `.eval()` to `.train()` will make the output are differenet? Why the output of pred are not the same in training and detecting processes?

**Comment by sezer-muhammed** [comment]
.eval and .train in general activate dropout layers.
I don't know if there are other layers activated or disabled.
In training process .train mode trains all layers with different data by dropping out some layers.
.eval mode activates all layers to show full performances. But maybe yolov5 don't have dropout layers who know

**Comment by LeongVan** [comment]
> 

There are no dropout layers, but do have some batch normalization layers, which will also activate in `.train()` and inactivate in `.eval()`. But will these activation make the probabilities of each class to become negative number?

**Comment by glenn-jocher** [comment]
See Detect layer
https://github.com/ultralytics/yolov5/blob/5a134e06530a8c24fdb9774c2c4ab0b513b08260/models/yolo.py#L77

**Comment by LeongVan** [comment]
> See Detect layer
> 
> https://github.com/ultralytics/yolov5/blob/5a134e06530a8c24fdb9774c2c4ab0b513b08260/models/yolo.py#L77

Thx!!! Now I see it~!

**Comment by xlnn** [comment]
> the rest of them are probabilities of all classes. Why probabilities are negative number?

Hello. I have the same the problem! The rest of them are probabilities of all classes. Why probabilities are negative number?
Do you know? Thank you!

**Comment by glenn-jocher** [comment]
@xlnn hello! Thank you for bringing up the issue.

In YOLOv5, the probabilities of all classes should not be negative numbers. Negative probabilities do not make sense in the context of object detection. 

There might be some other issue causing negative probabilities to be displayed. It could be a data preprocessing issue, a bug in the code, or some other factor. 

I recommend checking your data preprocessing steps to ensure that the input data is being properly normalized and scaled. Additionally, make sure that you are using the latest version of YOLOv5 and that you have followed the correct training and inference procedures.

If you are still experiencing this issue, please provide more details or code snippets that can help us reproduce the problem. This will allow us to better understand the issue and provide you with a more accurate solution.

Thank you!

**Comment by xlnn** [comment]
> @xlnn hello! Thank you for bringing up the issue.
> 
> In YOLOv5, the probabilities of all classes should not be negative numbers. Negative probabilities do not make sense in the context of object detection.
> 
> There might be some other issue causing negative probabilities to be displayed. It could be a data preprocessing issue, a bug in the code, or some other factor.
> 
> I recommend checking your data preprocessing steps to ensure that the input data is being properly normalized and scaled. Additionally, make sure that you are using the latest version of YOLOv5 and that you have followed the correct training and inference procedures.
> 
> If you are still experiencing this issue, please provide more details or code snippets that can help us reproduce the problem. This will allow us to better understand the issue and provide you with a more accurate solution.
> 
> Thank you!

Hello, I think is it possible that the use of the SiLu activation function in training is generating negative numbers. But even though it is generating negative numbers now, the detection is correct when running detect.py, so strange, I don't know why? As shown in the picture.
![image](https://github.com/ultralytics/yolov5/assets/57608826/d4d7e271-cef5-49c9-b452-5d9c648be106)

**Comment by glenn-jocher** [comment]
@xlnn hello, thank you for sharing your findings.

It is indeed possible that the use of the SiLU activation function in training could generate negative numbers. The SiLU function, also known as the sigmoid-weighted linear unit, can produce negative values when its input is negative. However, negative values should not affect the detection results.

In the image you shared, it seems that despite the negative probabilities, the detection results are correct when running `detect.py`. This could be due to the specific thresholding mechanism used in the detection script, which determines whether a detection is considered valid based on a probability threshold.

It's important to note that during training, YOLOv5 uses the binary cross-entropy loss function, which does not enforce the predicted probabilities to be within a specific range. Therefore, negative probabilities might be encountered during the training process.

If you are concerned about the negative probabilities, you could try applying a softmax function to the predicted class probabilities during inference, which would ensure that the probabilities are positive and sum up to 1. However, please note that this step is not necessary for the detection script to work correctly.

If you have any other questions or concerns, please feel free to provide more details or code snippets that can help us further investigate the issue.

Thank you!

**Comment by xlnn** [comment]
> @xlnn hello, thank you for sharing your findings.
> 
> It is indeed possible that the use of the SiLU activation function in training could generate negative numbers. The SiLU function, also known as the sigmoid-weighted linear unit, can produce negative values when its input is negative. However, negative values should not affect the detection results.
> 
> In the image you shared, it seems that despite the negative probabilities, the detection results are correct when running `detect.py`. This could be due to the specific thresholding mechanism used in the detection script, which determines whether a detection is considered valid based on a probability threshold.
> 
> It's important to note that during training, YOLOv5 uses the binary cross-entropy loss function, which does not enforce the predicted probabilities to be within a specific range. Therefore, negative probabilities might be encountered during the training process.
> 
> If you are concerned about the negative probabilities, you could try applying a softmax function to the predicted class probabilities during inference, which would ensure that the probabilities are positive and sum up to 1. However, please note that this step is not necessary for the detection script to work correctly.
> 
> If you have any other questions or concerns, please feel free to provide more details or code snippets that can help us further investigate the issue.
> 
> Thank you!

Thank you for your answer！

**Comment by glenn-jocher** [comment]
@xlnn thank you for bringing up this issue!

Regarding the negative probabilities generated during inference with YOLOv5, it is indeed possible for the SiLU activation function to produce negative values when its input is negative. However, these negative values should not affect the detection results.

From the image you provided, it seems that despite the negative probabilities, the detection results are correct when using the `detect.py` script. This can be attributed to the specific thresholding mechanism employed in the detection script, which determines whether a detection is valid based on a probability threshold.

During training, YOLOv5 utilizes the binary cross-entropy loss function, which does not enforce predicted probabilities to be within a specific range. As a result, encountering negative probabilities during training is not uncommon.

If you have concerns regarding the negative probabilities, you can consider applying a softmax function to the predicted class probabilities during inference. This will ensure that the probabilities are positive and sum up to 1. However, it is important to note that this step is not necessary for the detection script to function correctly.

If you have any further questions or require additional guidance, please provide more details or code snippets so that we can better understand and address the issue.

Thank you for your contribution!
